Attribute certificate (RFC 5755) WIP

dhobsong opened the pull request:
Closes #14648 

Please see the checklist below for sub-task completion status.

Right now I've split this implementation over several patches to try to keep the sizes a little more manageable,
but I can squash them if that makes more sense.

Some of the areas of concern regarding this patch series are listed below.  Any feedback related to these (or anything else)
would be greatly appreciated

### Attribute values

RFC 5755 defines several Attribute types that have ASN.1 SEQUENCEs as their value fields. E.g. [group](https://datatracker.ietf.org/doc/html/rfc5755#section-4.4.4) and [role](https://datatracker.ietf.org/doc/html/rfc5755#section-4.4.5).   As far as I can tell, there is no nice way to handle Attributes that don't have string (or similar) values for printing or parsing.  The current code just uses the output of `ANS1_item_print()` for printing, and parses from the config file using `ASN1_generate_nconf()`.
Maybe some framework, similar to the `X509V3_EXT_METHOD` used for extensions might be useful for X509_ATTRIBUTEs?

### public API

RFC 5755 defines the attribute certificate format using several new ASN.1 sequence types that don't seem to be used in any other profiles.  I've tried to find a good balance of what to expose in the public API, but if there are any opinions on whether I should be exposing more or less, I would be happy to hear them.

### file separation
I'm not 100% sure that all of the function implementations are in the right location (correct function in correct file). I've tried to 
follow the way that the X509, X509_REQ and X509_CRL implementation is divided, but maybe there is some room for improvement there.

<!--
Thank you for your pull request. Please review these requirements:

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING.md

Other than that, provide a description above this comment if there isn't one already

If this fixes a GitHub issue, make sure to have a line saying 'Fixes #XXXX' (without quotes) in the commit message.
-->

#### Checklist
- [x] file / BIO input output for attribute certificates (AC)
- [x] setters / getters for AC fields
- [x] read AC attribute values from config file
- [ ] documentation is added or updated
- [ ] tests are added or updated

mattcaswell commented:
Nice!

Note that this is too late for inclusion in 3.0 and has been added to the "post 3.0" milestone. For that reason its unlikely to get much developer attention until 3.0 is out.

Some of the CI failures seem relevant.

dhobsong commented:
Hi @mattcaswell,

Thanks for the info.  I'll address the CI failures and push forward for now.

FdaSilvaYY commented on include/openssl/x509.h.in:
```diff
@@ -1076,6 +1085,84 @@ int X509_PUBKEY_get0_param(ASN1_OBJECT **ppkalg,
+const X509_NAME *X509_ACERT_get0_issuerName(const X509_ACERT *x);
+long X509_ACERT_get_version(const X509_ACERT *x);
+void X509_ACERT_get0_signature(const X509_ACERT *x,
+         const ASN1_BIT_STRING **psig, const X509_ALGOR **palg);
```
nit : align the `const` keyword of this declaration.

FdaSilvaYY commented on include/openssl/x509.h.in:
```diff
@@ -1076,6 +1085,84 @@ int X509_PUBKEY_get0_param(ASN1_OBJECT **ppkalg,
+const STACK_OF(X509_EXTENSION) *X509_ACERT_get0_extensions(const X509_ACERT *x);
+int X509_ACERT_print(BIO *bp, X509_ACERT *x);
+int X509_ACERT_print_ex(BIO *bp, X509_ACERT *x, unsigned long nmflags,
+                      unsigned long cflag);
```
nit : please align the arguments between the two lines like this :
```
int X509_ACERT_print_ex(BIO *bp, X509_ACERT *x, unsigned long nmflags,
                        unsigned long cflag);
```
(not sure of the display result of my example.)

FdaSilvaYY commented on crypto/x509/build.info:
```diff
@@ -14,7 +14,8 @@ SOURCE[../../libcrypto]=\
         v3_pcia.c v3_pci.c v3_ist.c \
         pcy_cache.c pcy_node.c pcy_data.c pcy_map.c pcy_tree.c pcy_lib.c \
-        v3_asid.c v3_addr.c v3_tlsf.c v3_admis.c
+        v3_asid.c v3_addr.c v3_tlsf.c v3_admis.c x509_acert.c t_acert.c \
```
suggestion: Declare your new files on its own line.
It will ease any future merge.

FdaSilvaYY commented on crypto/x509/t_acert.c:
```diff
@@ -0,0 +1,235 @@
+        mlch = '\n';
+    }
+
+    if (!(cflag & X509_FLAG_NO_HEADER)) {
```
nit : `    if ((cflag & X509_FLAG_NO_HEADER) != 0) {`
And so on...

dhobsong replied:
`if ((cflag & X509_FLAG_NO_HEADER) == 0) {`  I think.
The checks in t_x509.c, and t_req.c all use the `if (!(cflag & XXX))`  format, so
I thought it was better to keep consistent with the existing code (albeit in a different file).

FdaSilvaYY replied:
The Coding GuideStyle is to explicitly test this binary bit-fields:
`if ((cflag & X509_FLAG_NO_HEADER) == 0) {` , you're right

dhobsong replied:
It does?  I'm looking at the style guide at https://www.openssl.org/policies/codingstyle.html and the closest I can find
is Chapter 21, which says that `if (! (...) ) ` shouldn't be used for pointer checks and str...cmp functions.  
I can't seem to find anything about binary bit fields.  Am I looking at the wrong guide? or just the wrong place in the right guide? :)

FdaSilvaYY replied:
See #10725 , which try to grab and extend the current codingstyle.

dhobsong replied:
Great, thanks.  This is really helpful.  I'll look it over the PR again with these points in mind.

FdaSilvaYY commented on crypto/x509/t_acert.c:
```diff
@@ -0,0 +1,235 @@
+        if (BIO_write(bp, "\n", 1) <= 0)
+            goto err;
+
+        if ((holder_bcid = X509_ACERT_get0_holder_baseCertId(x)))
```
nit:         `if ((holder_bcid = X509_ACERT_get0_holder_baseCertId(x)) != NULL)`

FdaSilvaYY commented on crypto/x509/t_acert.c:
```diff
@@ -0,0 +1,235 @@
+    }
+    if (!(cflag & X509_FLAG_NO_EXTENSIONS)) {
+        exts = X509_ACERT_get0_extensions(x);
+        if (exts) {
```
nit: explicit test against NULL :
`if (exts != NULL) {`

FdaSilvaYY commented on crypto/x509/x509_acert.c:
```diff
@@ -0,0 +1,301 @@
+                              int strtype, X509_ACERT *acert)
+{
+    int ret = 0, i;
+    STACK_OF(CONF_VALUE) *attr_sk = NULL;
```
nit : useless NULL init.

FdaSilvaYY commented on crypto/x509/x509aset.c:
```diff
@@ -0,0 +1,150 @@
+
+    local = ASN1_STRING_dup(src);
+    if (local == NULL)
+        return 0;
```
push an OUT_OF_MEMORY error here, with a block like this :
```
 oom:
    ERR_raise(ERR_LIB_X509, ERR_R_MALLOC_FAILURE);
```

FdaSilvaYY commented on crypto/x509/x509aset.c:
```diff
@@ -0,0 +1,150 @@
+{
+    GENERAL_NAME *gen_name = NULL;
+    STACK_OF(GENERAL_NAME) *new_names = NULL;
+    X509_NAME *name_copy = NULL;
```
useless NULL init

FdaSilvaYY commented on crypto/x509/x509aset.c:
```diff
@@ -0,0 +1,150 @@
+
+    /* only v2Form is supported, so always create that version */
+    if (v2Form == NULL) {
+        v2Form = ACERT_ISSUER_V2FORM_new();
```
missing alloc failure test

FdaSilvaYY reviewed (commented):
Quick review .
Very good work !
Mostly a bunch of nits.
Memory allocation failure handling need some rework in a few places.

dhobsong commented:
@FdaSilvaYY, thanks for the quick review.  
I'll make sure to add in the missing oom error blocks and address the nits and make sure there are no other instances of the same (there probably are).
I switched my compiler to clang from gcc, which highlighted one or two more warnings, so I'll address those too.

Any comments about how to print X509_ATTRIBUTEs in a generic way?  Maybe I should open a separate issue for that.

FdaSilvaYY commented on crypto/x509/x509_acert.c:
```diff
@@ -0,0 +1,307 @@
+    return x->acinfo->holder.baseCertificateID;
+}
+
+OBJECT_DIGEST_INFO * X509_ACERT_get0_holder_digest(const X509_ACERT *x)
```
nit: `OBJECT_DIGEST_INFO *X509_ACERT_get0_holder_digest(...`

FdaSilvaYY commented on crypto/x509/x_ietfatt.c:
```diff
@@ -0,0 +1,156 @@
+    }
+
+    val = IETF_ATTR_SYNTAX_VALUE_new();
+
```
missing alloc failure checking

FdaSilvaYY commented on crypto/x509/x_ietfatt.c:
```diff
@@ -0,0 +1,156 @@
+        }
+    }
+
+    for (int i = 0; i < IETF_ATTR_SYNTAX_get_value_num(a); i++) {
```
nit : `for (int i = 0; ...  ` is not compliant  with old _C89_ standard (not sure about C<number)>.
This counter variable declaration should be moved to the beginning of the method.

FdaSilvaYY reviewed (commented):
A few more nits ;)
All my previous comments  seem to be fixed, please feel free to resolve them ( just hit the resolve button)

dhobsong commented:
## v2 changes
* Addressed the review comments from @FdaSilvaYY  (thank you again for your feedback)
* Changed the argument / return type  of  `X509_ACERT_get0/set0_holder_entityName()` from X509_NAME to GENERAL_NAMES
   *  reason: entity name needs to be able to match against a value the SAN in the holder certificate, which can contain types other than X509_NAME
* Moved all getter/setter function definitions from x509.h.in to x509v3.h.in
    * reason:  `X509_ACERT_get0/set0_holder_entityName()` needs the definition of GENERAL_NAME which is only defined in x509v3.h.in.  Other functions may relay on other "V3 only" definitions in future.

FdaSilvaYY commented on crypto/x509/x_ietfatt.c:
```diff
@@ -0,0 +1,161 @@
+        break;
+    }
+
+    if (sk_IETF_ATTR_SYNTAX_VALUE_push(a->values, val) <= 0)
```
`val` should be free, on `sk_IETF_ATTR_SYNTAX_VALUE_push` failure.

FdaSilvaYY commented on crypto/x509/x_ietfatt.c:
```diff
@@ -0,0 +1,161 @@
+int IETF_ATTR_SYNTAX_print_ex(BIO *bp, IETF_ATTR_SYNTAX *a, int indent)
+{
+    int i;
+    if (a->policyAuthority != NULL) {
```
missing blank line here

FdaSilvaYY commented on include/crypto/x509.h:
```diff
@@ -307,6 +307,58 @@ struct x509_object_st {
     } data;
 };
 
+/* Attribute certificate defiinitions */
```
Typo `defiinitions` here

FdaSilvaYY commented on crypto/x509/x_ietfatt.c:
```diff
@@ -0,0 +1,161 @@
+    STACK_OF(IETF_ATTR_SYNTAX_VALUE) *values;
+};
+
+ASN1_CHOICE(IETF_ATTR_SYNTAX_VALUE) = {
```
Some ASN.1 structure definitons are commented with a copy-paste of ASN.1 definition from RFC.
I may find some example, if you don't get them ?

FdaSilvaYY commented on crypto/x509/x509_acert.c:
```diff
@@ -0,0 +1,307 @@
+    ret = 1;
+err:
+    return ret;
+
```
nit: useless blank line ...

FdaSilvaYY commented on crypto/x509/x509_acert.c:
```diff
@@ -0,0 +1,307 @@
+#include <openssl/x509v3.h>
+#include "x509_acert.h"
+
+ASN1_SEQUENCE(OBJECT_DIGEST_INFO) = {
```
May be a copy of ASN.1 structure definitions from RFC could be a good comment ?

dhobsong replied:
The structure definition for OBJECT_DIGEST_INFO, together with the other structure defined in this file
gets to be fairly long.  I've added a comment referencing the RFC and section number as is done in other
places, like `crypto/x509/v3_addr.c`

FdaSilvaYY reviewed (commented):
A very few remaining nits.
Seem almost good for me.

dhobsong commented:
## v3 changes

* Addressed previous review comments
* Make sure getter / setter functions that handle pointer types use the get0/get1 set0/set1 convention
* Added attribute certificate serial number to the output of X509_ACERT_print_{ex} functions
* renamed `IETF_ATTR_SYNTAX_print_ex()` to `IETF_ATTR_SYNTAX_print()` as no previous version exists

JonathanWilbur commented:
@dhobsong Can I ask you what your expectations are for the scope of your attribute certificates implementation? Are you expecting to implement:

- Signature verification?
- `Validity` checking?
- `holder` matching?
- ACRL checking?
- OCSP checking?
- Any particular extensions?
- Any particular attributes?
- Command-line usage?

Thank you for starting this!

dhobsong commented:
Hi @JonathanWilbur 
I was looking to implement
* Signature verification
* Validity checking (i.e. the checks from Section 5 of the RFC)
* holder matching (if you mean verifying that a given attribute certificate corresponds to a given PKI certificate)
* Command-line usage

The current patch set implements the IETFAttrSyntax, but I may also use some others (not yet decided).  As I mentioned in the top of this PR though, some more work may be needed to better handle setting/parsing before adding more new attributes.

TBH I hadn't really give OCSP, ACRL and additional extensions much thought so far.

All of this is dependent on this PR of course though.

dhobsong commented:
## v4 changes
* Added / updated documentation 
* made `X509_ACERT_get0_issuerName()` and `X509_ACERT_get0_info_signature()` return non-const values to make them similar to the `X509` versions
* Added `X509_ACERT_get_signature_nid()` to be consistent with the signature API for `X509`, `X509_CRL`, etc.

The CI test suite should now pass all tests (previously was failing doc-nits), so I will remove the draft status.  Comments are still very welcome.

FdaSilvaYY commented on doc/man3/IETF_ATTR_SYNTAX.pod:
```diff
@@ -0,0 +1,86 @@
+
+=item IETFAS_OCTECTS
+
+A pointer to an ASN1_OCTECT_STRING
```
Typo: type is ASN1_OCTET_STRING !

FdaSilvaYY commented on doc/man3/IETF_ATTR_SYNTAX.pod:
```diff
@@ -0,0 +1,86 @@
+
+=over 4
+
+=item IETFAS_OCTECTS
```
IETFAS_OCTE_**C**_TS ?

FdaSilvaYY commented on doc/man3/X509V3_get_d2i.pod:
```diff
@@ -71,9 +72,10 @@ X509_REVOKED_get_ext_d2i() and X509_REVOKED_add1_ext_i2d() operate on the
-of a certificate a CRL or a CRL entry respectively.
+X509_get0_extensions(), X509_CRL_get0_extensions(),
+X509_ACERT_get0_extensions(), and X509_REVOKED_get0_extensions() return a
+stack of all the extensions of a certificate a CRL an attribute certificaate
```
weird sentence , missing commas ? 
Typo : certificaate

FdaSilvaYY commented on doc/man3/X509_get_subject_name.pod:
```diff
@@ -59,22 +63,29 @@ X509_get_subject_name(), X509_set_subject_name(), and X509_subject_name_hash()
 or issuer names of certificate requests of CRLs respectively.
 
+Since attribute certificates do not have a subject name, only the issuer name
+can be set.  For details on setting X509_ACERT holder identies, see
```
typo: identies

FdaSilvaYY reviewed (commented):
quick review of the doc mostly

dhobsong commented:
rebase and re-push

JonathanWilbur commented:
I see that this task is in OMC review. I know not a lot of people don't know about attribute certificates and their significance. I'd be happy to speak to the importance of attribute certificates if anybody in the OMC feels they could use some guidance on what attribute certificates are and how attribute certificates are used.

t-j-h commented:
Post a summary of your views here. That helps everyone see the context you are focused on.

JonathanWilbur commented:
X.509 certificates got their name because they originate from [ITU Recommendation X.509](https://www.itu.int/rec/T-REC-X.509-201610-S) (not IETF RFC 5280 or its precursors). But this is not the only thing specified in ITU Recommendation X.509. This standard also defines attribute certificates, which function more like a JSON Web Token: they are a general-purpose digitally authenticated collection of attributes (although their intended use was primarily privilege management). The equivalent of the public-key certificates `subject` field in an attribute certificate is the `holder` field, which relates the group of attributes to a public-key certificate, to an X.500 directory name, or something else.

Several large companies (I am under a non-disclosure agreement, but you'd recognize their names) are interested in seeing attribute certificate support incorporated into OpenSSL. At least one such sought-after use case I know of is Platform Certificates (a subtype of an attribute certificate), which provide supply chain integrity for hardware. I'd be happy to delve into further detail, but here are some links pertaining to platform certificates and attribute certificates that might explain this better than I could.

- https://www.youtube.com/watch?v=Yh-OTPXBaps
- https://www.youtube.com/watch?v=muKP_2NMSFo
- https://www.nccoe.nist.gov/supply-chain-assurance
- https://en.wikipedia.org/wiki/Authorization_certificate

dhobsong commented:
Thanks @JonathanWilbur for advocating for this, and for adding relevant reference material.

Just to add a little bit of context, my particular use case is for privilege management, which is why this
patch series includes support for the IETFAttrSyntax type, which is defined in the spec for representing group membership of the attribute certificate holder.

I've rebased the patch series again.  The CI tests all pass on my local branch. The last time I pushed, it seemed like there was
and issue with AVX512 support unrelated to this patch series.

tatac1 commented:
Thank you @JonathanWilbur for advocating for this,  Actually I am working with @dhobsong.

RFC 5755 is used [TCG Platform Certificate Profile](https://trustedcomputinggroup.org/wp-content/uploads/IWG_Platform_Certificate_Profile_v1p1_r19_pub_fixed.pdf). However, as far as I know, there is only one software that supports the specification, provided by the NSA.

https://github.com/nsacyber/paccor

There is a lot of talk about the importance of the supply chain(i.e SP 800-147, SP 800-193), but  support tools is limited.

RFC 5755 is to make it possible for OpenSSL, which is used in a variety of environments, users can use OpenSSL to verify the firmware of their devices.

AndreasFuchsSIT commented:
+1 on this feature.
I've hacked around TCG Platform Certificates in the past but it was no fun due to lack of broader support.
Platform Certs as such are a great concept but suffer from chicken-egg problem. So adoption by OpenSSL as major software library would be a great enabler.
Also from the [tpm2-software/tpm2-tss](https://www.github.com/tpm2-software/tpm2-tss) standpoint we have the platform cert retrieval functions in place.

(Full disclosure: I'm also a TCG member; focused on the TPM Software Stack)

williamcroberts commented:
+1 on he feature, I work on the tpm2-software org here on github, and it would be nice to have this in place in OSSL.

openssl-machine commented:
This PR is in a state where it requires action by @openssl/omc but the last update was 39 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/omc but the last update was 70 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/omc but the last update was 101 days ago

mattcaswell commented:
OMC have voted to lift the OMC hold on this PR.

t-j-h commented:
You will need to rebase the PR to the fresh master branch to resolve conflicts.

dhobsong commented:
> You will need to rebase the PR to the fresh master branch to resolve conflicts.

@t-j-h, thanks for the heads up.  I'm moving house this week, but I think that I should be able to rebase and retest this late next week, or the week following.

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 30 days ago

JonathanWilbur commented:
@FdaSilvaYY and @iqbal0kabat , is there any reason this couldn't be merged in now? I have another [pull request](https://github.com/openssl/openssl/pull/16607) outstanding that would provide no added functionality unless this were merged in.

FdaSilvaYY commented:
This PR needs to be reviewed and approved by @openssl members. 
Subject to any merge conflict already resolved.
I'm just trying to reduce the team burden by reviewing a bit some changes, when I have some spare time.

t8m commented on include/openssl/x509.h.in:
```diff
@@ -281,6 +281,11 @@ typedef struct SCRYPT_PARAMS_st {
 } SCRYPT_PARAMS;
 #endif
 
+typedef struct x509ACert_st X509_ACERT;
+typedef struct x509ACertInfo_st X509_ACERT_INFO;
+typedef struct object_digest_info_st OBJECT_DIGEST_INFO;
+typedef struct IssuerSerial_st ISSUER_SERIAL;
```
These identifiers and also the IETF_ATTR_SYNTAX* identifiers break the rule that new identifiers should be prefixed with ossl_ or OSSL_ prefix. Unfortunately it would make the identifiers even longer than they currently are. I suppose we want OTC to talk about this.

t8m commented:
**OTC: Are the newly added public identifiers OK, or do they have to be prefixed with OSSL_/ossl_ prefix?**

mspncp commented:
IIRC, the OSSL_/ossl_ prefix policy applies strictly only to new 'namespaces', not to existing ones like 'x509'.

In the case of `x509.h` it would be quite inconsistent to start adding those prefixes.

t8m commented:
@mspncp please note there are identifiers like: OBJECT_DIGEST_INFO, ISSUER_SERIAL, IETF_ATTR_SYNTAX being added.

t8m commented:
Maybe one option would be to add a completely new header file instead of adding the attribute certificate related stuff into x509.h and x509v3.h.

dhobsong commented:
@t8m 
> @mspncp please note there are identifiers like: OBJECT_DIGEST_INFO, ISSUER_SERIAL, IETF_ATTR_SYNTAX being added.

Just to clarify, would the identifiers under the X509_ namespace (ie. X509_ACERT, X509_ACERT_INFO) be fine as is?

dhobsong commented:
Right now, OBJECT_DIGEST_INFO and the like have been named so that there are consistent with the extensions defined in the same file such as NAMING_AUTHORITY.  Having a separate header for IETF_ATTR_SYNTAX and future attributes/extensions might be useful.  Would that be only for attribute certificate stuff though? Maybe definitions for other extension types could end up there too.

t8m commented:
> Just to clarify, would the identifiers under the X509_ namespace (ie. X509_ACERT, X509_ACERT_INFO) be fine as is?

IMO these are fine.

t8m commented:
**OTC Decision: The OBJECT_DIGEST_INFO*, ISSUER_SERIAL* and IETF* namespaces are not ours to
claim. All such names should be prefixed with OSSL_.**

dhobsong commented:
@t8m Thank you for the update. I'll move those definitions out to a separate header and add the OSSL_ prefix.

dhobsong commented:
@t8m I've moved all of the attribute certificate symbols and API definition to a new header 'x509_acert.h', updated the documentation to reflect this and rebased to the latest master.

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 30 days ago

JonathanWilbur commented:
@t8m Are you (or other OpenSSL reviewers) waiting on these merge conflicts to be resolved to review this PR?

dhobsong commented:
@t8m I've rebased and addressed the merge conflicts.  
I'll try to be more vigilant with rebases in future to avoid holding up the review process. 
Is there anything else blocking this review right now?

dhobsong commented:
Oops, sorry. It looks like I broke the CI build. I'll address it and re-push

FdaSilvaYY reviewed (commented):
Uncomplete review

FdaSilvaYY commented on util/missingcrypto.txt:
```diff
@@ -672,6 +672,10 @@ IDEA_ofb64_encrypt(3)
 IDEA_options(3)
 IDEA_set_decrypt_key(3)
 IDEA_set_encrypt_key(3)
+OSSL_IETF_ATTR_SYNTAX_VALUE_free(3)
```
This is a merge blocking change.
This file is the undocumented methods inventory.
It should not be extended.

There is a dedicated pod file to document these ASN1 framework-generated methods.

dhobsong replied:
man3/X509_dup.pod?  I'll move OSSL_IETF_ATTR_SYNTAX_VALUE_{free, it, new} there. (ACERT_INFO too)

What about the various d2i, i2d and _print functions? I've put those in this file because similar functions for the other X509 types are also undocumented.

dhobsong replied:
@FdaSilvaYY Thanks again for looking at this.

I've added documentation for most of the new API functions. The exceptions are the 
`X509_ACERT_print()` functions (which have a similar API to the `X509_print()` functions) and 
`X509_ACERT_add_attr_nconf()` (similar to `X509_ACERT_add_attr_nconf()`), which are currently undocumented.  

It's probably useful to add documentation for all of these versions (newly added and existing), but maybe in a separate PR?

dhobsong commented:
The CI build failures are caused by a macro name conflict between a recently added [test case](https://github.com/openssl/openssl/blob/master/test/ext_internal_test.c#L17) and the [x509v3.h header](https://github.com/openssl/openssl/blob/master/include/openssl/x509v3.h.in#L397) which this MR adds to the files included by [pem.h](https://github.com/openssl/openssl/blob/e99e44f39fd607b07b01478d3c02490893ad0876/include/openssl/pem.h#L24).  

I can rework this series to call the {DECLARE,IMPLEMENT}_PEM* macros directly from the x509_acert.h/x509_acert.c which avoids having to add all the x509v3.h symbols to the namespace when including pem.h.  The only other types that are doing something like this are CMS and SSL_SESSION, so I'm not sure if there are any reasons to avoid doing this.  Are there any thoughts or comments on this approach?

t8m commented:
The x509v3.h header file is really ugly in terms of invading other namespaces. It should not be added to any of the other existing includes. If you need to include it in x509_acert.h it can be acceptable as this is a new file.

dhobsong commented:
Thanks.  That lines up with what I was thinking.

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 30 days ago

dhobsong commented:
@t8m I've rebased this onto master again.  Please let me know if there is anything else on my end blocking progress on this.

t8m commented:
> @t8m I've rebased this onto master again. Please let me know if there is anything else on my end blocking progress on this.

Thank you, this is currently blocked on finding reviewers with enough free time from the otc/committers as this is a quite big pull request.

t8m commented on util/missingcrypto.txt:
```diff
@@ -1124,6 +1124,9 @@ X509V3_section_free(3)
 X509V3_set_conf_lhash(3)
 X509V3_set_nconf(3)
 X509V3_string_free(3)
+X509_ACERT_add_attr_nconf(3)
+X509_ACERT_print(3)
+X509_ACERT_print_ex(3)
```
This is unfortunately not allowed. We do not allow any additions to missingcrypto.txt. You will need to add at least some documentation.

t8m commented:
One more thing that will be probably required before we can merge this is more testing. I.e. some API tests for the newly added APIs and fuzz testing of parsing of the attribute certificates.

dhobsong commented:
@t8m Thank you for the comments. I'll look into addressing those points.

FdaSilvaYY reviewed (commented):
a few more nits.

JonathanWilbur commented:
Hey there, checking in on this one. Can I help at all?

Jr-produciones1000 commented:
saludos cordiales rss.jr.producciones1000.xml<https://drive.google.com/uc?id=14C72-S9oB30juDSvd7grMLlHBPUQ5imt&export=download>

Obtener Outlook para Android<https://aka.ms/AAb9ysg>
________________________________
From: Jonathan Wilbur ***@***.***>
Sent: Sunday, December 25, 2022 8:32:36 AM
To: openssl/openssl ***@***.***>
Cc: Subscribed ***@***.***>
Subject: Re: [openssl/openssl] Attribute certificate (RFC 5755) (#15857)


Hey there, checking in on this one. Can I help at all?

—
Reply to this email directly, view it on GitHub<https://github.com/openssl/openssl/pull/15857#issuecomment-1364691104>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/A35B5NMFREODASSSYE5W7SDWPBLIJANCNFSM47DDE33Q>.
You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>

FdaSilvaYY commented on crypto/x509/x509aset.c:
```diff
@@ -0,0 +1,173 @@
+
+    s = ASN1_STRING_dup(src);
+    if (s == NULL) {
+        ERR_raise(ERR_LIB_X509, ERR_R_ASN1_LIB);
```
See #19301, it will reduce your added code length.

dhobsong replied:
Thanks,  I'll get rid of these redundant calls.

dhobsong replied:
Looking at #19301, it seems like the `ERR_raise` calls are removed only in places that call `OPENSSL_malloc()` directly.  Indirect calls, through the ASN1 library for example, just update the failure reason, which is what the current code is already doing.

FdaSilvaYY commented:
> Hey there, checking in on this one. Can I help at all?

@JonathanWilbur may I suggest to write a few more test code, or test samples ?
There is a test who compare x509 dump output to reference file.

@dhobsong  : you may run the fuzz code on your new declared structure. see `fuzz.c`

Jr-produciones1000 commented:
https://youtube.com/shorts/crKoPZK_ya8?feature=share

Oscar Gonzalez Lara

JonathanWilbur commented:
@FdaSilvaYY What would be needed in terms of tests? I can produce attribute certificates via other tools for testing purposes. I just need to know what you would be looking for. It looks like there is already [a test](https://github.com/openssl/openssl/blob/fa3a322d109e6b7b9230115528d08842fb487e52/test/recipes/60-test_x509acert.t). Is there something that this test lacks specifically?

JonathanWilbur commented:
Hey @dhobsong have you given up on this PR? Not shaming you; I know it's been hard. I just want to know if I should continue developing this.

FdaSilvaYY commented:
> @FdaSilvaYY What would be needed in terms of tests? I can produce attribute certificates via other tools for testing purposes. I just need to know what you would be looking for. It looks like there is already [a test](https://github.com/openssl/openssl/blob/fa3a322d109e6b7b9230115528d08842fb487e52/test/recipes/60-test_x509acert.t). Is there something that this test lacks specifically?

I remember seeing a test where the cert or key output dump/log from an `OpenSSL` command is redirected to a text file.
And this output text file is compared to a reference file to detect any parsing/decoding alteration.

dhobsong commented:
Hi @JonathanWilbur, I haven't given up on the PR, but I've just had too much on my plate to follow up with it recently.  I was looking at the documentation side, but I haven't had a chance to look at any additional testing at all. If you're interested in providing some additional tests, I'd very much welcome that.

dhobsong commented:
@FdaSilvaYY The openssl command was not included in this PR in the interest of keeping it shor(er), so I guess that the equivalent would be a test that generates an attribute certificate with the API and writes that out to compare against a golden file.

JonathanWilbur commented:
For the reasons stated above, I don't think this could be tested the way you described @FdaSilvaYY . Would you want the command-line app code included in this PR so that it could be tested? Or what if I just created a fork of this repo that included an app that creates an attribute certificate? Would that be sufficient proof of it's validity?

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 30 days ago

dhobsong commented:
@JonathanWilbur I've added some fuzz testing for attribute certificate parsing (not pushed yet), but I could use some more sample data to diversify the corpus.  You mentioned previously that you have other tools for generating attribute certificates.  Do you have some samples of valid and invalid certificates that you would be willing to contribute?

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 30 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 61 days ago

JonathanWilbur commented:
Hey @dhobsong sorry I didn't realize you were waiting on me for an answer. I created a PR containing 19 new attribute certificates: https://github.com/openssl/openssl/pull/21231

dhobsong commented:
Hi @JonathanWilbur, thanks for the certificates.  I created a few test certificates of my own that I was able to use for some initial fuzz testing, so I've already started with that.  
More test data is always better though.

JonathanWilbur commented:
@dhobsong they had apprehension about accepting the certs in that PR without CLAs from all the original authors, so I closed that PR. I had doubts about even making it. I will produce some attribute certificates using my own tooling, instead, then.

JonathanWilbur commented:
@dhobsong I created an attribute certificate using [my own tooling](https://github.com/Wildboar-Software/directory/pull/19). I wrote the ASN.1 library, X.500 library, X.500 directory server, and the feature code to create attribute certificates, so it is _very_ mine, and I am explicitly giving you any rights necessary to use this attribute certificate in your PR:

```
-----BEGIN ATTRIBUTE CERTIFICATE-----
MIIDVzCCAT8CAQEwE6ERpA8wDTELMAkGA1UEBhMCU06ggcIwU6RRME8xCzAJBgNV
BAYTAlVTMQswCQYDVQQIDAJGTDEOMAwGA1UEBwwFVGFtcGExETAPBgNVBAoMCFdp
bGRib2FyMRAwDgYDVQQDDAdtZWVya2F0oGswU6RRME8xCzAJBgNVBAYTAlVTMQsw
CQYDVQQIDAJGTDEOMAwGA1UEBwwFVGFtcGExETAPBgNVBAoMCFdpbGRib2FyMRAw
DgYDVQQDDAdtZWVya2F0AhQaaWjvGVDkF3Tx4GAP4wPbxfNUyjANBgkqhkiG9w0B
AQ0FAAIQZqnqtEsmzSO+smBA0hjP7zAiGA8yMDIzMDYyNzE2MTcwOVoYDzIwMjMw
NjI3MTcxNzA5WjAbMAsGA1UEBjEEEwJTTjAMBgNVBAAxBQYDVQYCMA0GCSqGSIb3
DQEBDQUAA4ICAQBtKz9tbdnoH8fscjSKN8sG30beZ2P5GGiwK6gt419vBgT2Gz7T
2d7qKTiSdu3Ti7RludHTUfAmrpsrBRdL47PlX64nbSOLE+n7WwdUU/EkKK8Qmu6m
xoUp5nzeqEiijHE4QIrtH4MfrxGBUawsRjPazyZxW/YzGNeC57VzV4BZGjghL2Ta
wNjIOyJiaoCudbYIIWPE2jIuR8DZ4cXe+SDY//gBUy4gYdLmJLqrctpwd6cOhxSn
HM5cDwFUTqREB6O82HrLxivdcbpZT5vpL7x437KSHK2TzCivXwaT+AONpP+cx+kJ
kTtNs36L8VaP7u2cCDC2e6SoI2wm4/erUp0UuBYpY8QZhpblL03OG4K7p/tWf9RB
cyJhqfqO/cLbSWSK1cB4oJJCr8uj81qVYpqiMXatP+lzFwrMRjorVnNwCJDVCLUG
CXqYRojqhFIIksDGaTa0aEqCyuaCVUaN7nnTZaULeK+8SmGcDNQLyjiXjrQJB5u3
Q24EA398LYEBRZ7JsWl9FR26SvE3XyC9OI7LUvAi5Bx/ng7YVVBXQttyYwwmUfCU
39RWWzPbu1WZ1CXv+jvmwLgGjHp4VURBSF6jHft31ZcmlM6iBc6qooMisfCYsPo/
6uZLH9htmWmHnWfBa7+ELBazRRpzKZhjvY72Nd3Qa3HJ9Xy1sCJDrR42mg==
-----END ATTRIBUTE CERTIFICATE-----
```

I can create more attribute certificates, if you're looking for anything in particular, but I figure one from a totally independent toolset is mostly what the maintainers want.

dhobsong commented:
Thanks so much. I can add this to my fuzz testing corpus and other test cases if necessary. 
Once the patch for that is ready, I will ask for your "Signed-off-by" to add to the commit.

dhobsong commented on crypto/x509/x509_acert.c:
```diff
@@ -0,0 +1,311 @@
+{
+    const char *p = *value;
+
+    if (strncmp(p, "ASN1:", 5) != 0)
```
`memcmp` isn't guaranteed to stop comparing early `p` is less than 5 (null-terminated) bytes.
 valgrind complains in that case, which is avoided with `strncmp`.

dhobsong commented:
I've added a new version that take into account the feedback that I have received so far.
Thank you again for the comments.

I'm not 100% sure of the merge procedure for the fuzz test corpus, though.  Does this need to get merged before the corpus file updates are?  

## v5 changes

* Added additional tests and fuzz test
* Added documentation for all public functions
* Rearranged commits so that implementation and documentation of functions are in the same commit
* Define getter/setter return values/parameter as `const` when appropriate
* Make sure all getters have setters and vice versa.
* Various minor updates to the attribute certificate printing functions

FdaSilvaYY commented on crypto/x509/x509_acert.c:
```diff
@@ -0,0 +1,329 @@
+}
+
+void OSSL_OBJECT_DIGEST_INFO_get0_digest(const OSSL_OBJECT_DIGEST_INFO *o,
+                                    int *digestedObjectType,
```
nit: align `int` with the `const` on above line.

FdaSilvaYY commented on crypto/x509/x509_acert.c:
```diff
@@ -0,0 +1,329 @@
+
+        if (value == NULL) {
+            ERR_raise_data(ERR_LIB_X509, X509_R_INVALID_ATTRIBUTES,
+                    "name=%s,section=%s",v->name, section);
```
nit: should be : 
```
            ERR_raise_data(ERR_LIB_X509, X509_R_INVALID_ATTRIBUTES,
                           "name=%s,section=%s",v->name, section);
```

FdaSilvaYY commented on crypto/x509/x509aset.c:
```diff
@@ -0,0 +1,177 @@
+}
+
+int OSSL_OBJECT_DIGEST_INFO_set1_digest(OSSL_OBJECT_DIGEST_INFO *o,
+                                         int digestedObjectType,
```
nit: one space less

FdaSilvaYY commented on crypto/x509/x509aset.c:
```diff
@@ -0,0 +1,177 @@
+{
+
+    if (ASN1_ENUMERATED_set(&o->digestedObjectType, digestedObjectType) <= 0)
+	    return 0;
```
nit: 4 space less

FdaSilvaYY commented on doc/man3/X509_ACERT_print_ex.pod:
```diff
@@ -0,0 +1,112 @@
+
+= item *
+
+The identity of the holder of the attribute certificatate. If the
```
typo: certificatate

FdaSilvaYY reviewed (commented):
only a few nits.
Look good !!

dhobsong commented:
> only a few nits. Look good !!

Thanks again for looking at this.  Nits addressed and re-pushed.

t8m commented:
> I'm not 100% sure of the merge procedure for the fuzz test corpus, though. Does this need to get merged before the corpus file updates are?
> ##

Adding the fuzz corpus should be a separate PR after this one gets merged, IMO.

JonathanWilbur commented:
@t8m @FdaSilvaYY Can the `hold: needs tests` tag be removed from this one now?

FdaSilvaYY commented on crypto/x509/x509_acert.c:
```diff
@@ -0,0 +1,329 @@
+{
+    STACK_OF(X509_ATTRIBUTE) **attrs = &x->acinfo->attributes;
+
+    return (X509at_add1_attr_by_txt(attrs, attrname, type, bytes, len) != NULL);
```
Nit: return without ()

FdaSilvaYY commented on crypto/x509/x509_acert.c:
```diff
@@ -0,0 +1,329 @@
+const X509_NAME *X509_ACERT_get0_issuerName(const X509_ACERT *x)
+{
+    if (x->acinfo->issuer.type != X509_ACERT_ISSUER_V2 ||
+        x->acinfo->issuer.u.v2Form == NULL)
```
Nit : || should be on this line

FdaSilvaYY commented on crypto/x509/t_acert.c:
```diff
@@ -0,0 +1,291 @@
+                            at->value.sequence->length, i, 1);
+            break;
+        default:
+            if (BIO_puts(bp, "unable to print attribute\n") <= 0)
```
Personally, I will add a 0x%X to print this unexpected type value

FdaSilvaYY commented on doc/man3/X509_ACERT_get0_holder_baseCertId.pod:
```diff
@@ -0,0 +1,119 @@
+ OSSL_OBJECT_DIGEST_INFO * X509_ACERT_get0_holder_digest(const X509_ACERT *x);
+ void X509_ACERT_set0_holder_entityName(X509_ACERT *x, GENERAL_NAMES *name);
+ void X509_ACERT_set0_holder_baseCertId(X509_ACERT *x, OSSL_ISSUER_SERIAL *isss);
+ void X509_ACERT_set0_holder_digest(X509_ACERT *x,
```
Nit: indent is 1 space off

FdaSilvaYY commented on doc/man3/X509_ACERT_get0_holder_baseCertId.pod:
```diff
@@ -0,0 +1,119 @@
+ int OSSL_ISSUER_SERIAL_set1_issuerUID(OSSL_ISSUER_SERIAL *isss, ASN1_BIT_STRING *uid);
+
+ void OSSL_OBJECT_DIGEST_INFO_get0_digest(OSSL_OBJECT_DIGEST_INFO *o,
+                                      ASN1_ENUMERATED **digestedObjectType,
```
Indent is off

FdaSilvaYY commented on doc/man3/X509_ACERT_print_ex.pod:
```diff
@@ -0,0 +1,112 @@
+The list of attributes contained in the attribute certificate.
+The attribute type is printed with i2a_ASN1_OBJECT().  String valued
+attributes are printed as raw string data. ASN1 encoded values are
+printed witth ANS1_parse_dump().  (X509_FLAG_NO_ATTRIBUTES)
```
Typo: witth

FdaSilvaYY commented on doc/man3/X509_ACERT_print_ex.pod:
```diff
@@ -0,0 +1,112 @@
+X509_NAME_print_ex(). See L<X509_NAME_print_ex(3)> for details.
+
+X509_ACERT_print() is equivalent to calling X509_ACERT_print_ex() with the
+I<nmflags> and I<cflags> set to to XN_FLAG_COMPAT and X509_FLAG_COMPAT
```
Typo: to to

FdaSilvaYY commented on doc/man3/X509_get0_uids.pod:
```diff
@@ -25,6 +32,9 @@ practice outside test cases.
 
 X509_get0_uids() does not return a value.
 
+X509_ACERT_get0_issuerUID() retuns a unique identifier on success or NULL
```
Typo: retuns

FdaSilvaYY commented on doc/man3/X509_get_serialNumber.pod:
```diff
@@ -28,12 +35,19 @@ X509_set_serialNumber() sets the serial number of certificate B<x> to
-X509_get_serialNumber() and X509_get0_serialNumber() return an B<ASN1_INTEGER>
-structure.
+X509_get_serialNumber(), X509_get0_serialNumber() and
+X509_ACERT_get0_serialNumber() return an B<ASN1_INTEGER> structure.
```
return a pointer to an ...

FdaSilvaYY commented:
A few nits remaining.
After that, it is good to go, IMHO

dhobsong commented:
@FdaSilvaYY Thanks for the review. I've addressed the nits and rebased to master.

FdaSilvaYY commented on crypto/x509/x509_acert.h:
```diff
@@ -0,0 +1,21 @@
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+#include <crypto/x509_acert.h>
```
personally, I will put this #include inside the guards, not outside.

dhobsong replied:
Ok. I can do that.

FdaSilvaYY commented on crypto/x509/x509aset.c:
```diff
@@ -0,0 +1,177 @@
+    }
+
+    if (sk_GENERAL_NAME_push(new_names, gen_name) <= 0) {
+        ERR_raise(ERR_LIB_X509, ERR_R_CRYPTO_LIB);
```
You may factorise the code in this method:
- only call `ERR_raise(ERR_LIB_X509, ERR_R_CRYPTO_LIB);` after end label
- it will spare some accolades.
- none of these point of failures are ASN1 related.

dhobsong replied:
`GENERAL_NAME_new` is defined by the `IMPLEMENT_ASN1_FUNCTIONS`.  Doesn't that make its failure ASN1 related?  That seems consistent with how the errors are handled in [several](https://github.com/openssl/openssl/blob/master/crypto/x509/v3_akid.c#L208) [other](https://github.com/openssl/openssl/blob/master/crypto/ess/ess_lib.c#L97) [places](https://github.com/openssl/openssl/blob/master/crypto/x509/v3_san.c#L452).

Maybe I'm not understanding something fundamental about these ERR reason codes.  I thought that they should explain the reason that the caller exited abnormally.  So if the caller calls an ASN1 related function and that function fails, ASN1 should be the reason code (unless there is a more specific reason).  Is that not the case?

> only call ERR_raise(ERR_LIB_X509, ERR_R_CRYPTO_LIB); after end label

If different failure conditions are raising different errors, then calling ERR_raise in each if block seems the correct thing to do.  If  `ERR_R_CRYPTO_LIB` is the correct code for all error cases in this function it can be refactored as you suggest.  That will depend on the resolution of the ASN1 error code question.

levitte replied:
> - only call `ERR_raise(ERR_LIB_X509, ERR_R_CRYPTO_LIB);` after end label

Er, why?  The source line number is registered at the `ERR_raise()` point, so conglomerating them to one call after `end:` will make tracing harder.

dhobsong replied:
Another good point.  

Are there any objections to resolving this comment, and leaving the code as-is?

t8m replied:
Resolving

FdaSilvaYY commented on crypto/x509/x509aset.c:
```diff
@@ -0,0 +1,177 @@
+    ASN1_BIT_STRING_free(isss->issuerUID);
+    isss->issuerUID = ASN1_STRING_dup(uid);
+    if (isss->issuerUID == NULL) {
+        ERR_raise(ERR_LIB_X509, ERR_R_ASN1_LIB);
```
not an ASN1 lib related error, just a crypto internal one here.

dhobsong replied:
Please see my comment above

dhobsong commented:
@FdaSilvaYY Thank you again for the comments.

With respect to `CHANGES.md` and `NEWS.md`, is this a good place to mention the things that are *not* included in this PR?
Two of the biggest missing pieces right now are:
- No command line support
- No nice mechanism for specifying/printing X509_ATTRIBUTES that are ASN1 sequences (currently using `ASN1_generate_nconf()` and `ANS1_item_print()`. See the first comment in this PR)

dhobsong commented:
@mattcaswell  should the "epic" tag also be applied to the original issue that this PR aims to address (in part) (#14648)?

mattcaswell commented:
> @mattcaswell should the "epic" tag also be applied to the original issue that this PR aims to address (in part) (https://github.com/openssl/openssl/issues/14648)?

Actually, it is probably better applied to the issue than this PR. I've moved it there.

JonathanWilbur commented:
@FdaSilvaYY I think @dhobsong has been waiting on your response.

dhobsong commented:
@FdaSilvaYY, As Jonathan says, I was hoping to get some response to my previous comments.  Could you let me know your thoughts?

@mattcaswell, @t8m Do you see any remaining significant blockers that have not yet been addressed in this PR?

FdaSilvaYY reviewed (commented):
Few small nits spotted in docs.
About reason code  in `ERR_raise(ERR_LIB_X509,  ERR_R_... )` calls, reviewers will say a word if needed.
Ping @openssl, this is ready for review.

@dhobsong  : sorry for my late return, daily job shores.
You may push your next packet of changes in a new PR, including this commits ?

dhobsong commented:
> sorry for my late return, daily job shores.

@FdaSilvaYY  No problem. I completely understand how busy things can get. Thank you again for the time that you have spent looking at this PR.

>About reason code in ERR_raise(ERR_LIB_X509,  ERR_R_... ) calls, reviewers will say a word if needed.

Ok.  I'll leave those calls as is for now, and invite the maintainers to comment if something need to be changed there.

I'll address those documentation comments and resubmit.

dhobsong commented:
I noticed that the CIFuzz fuzzing test has failed since the acert corpus has not yet been merged into the fuzz-corpora repo.  
Ⅰ was planning to submit a PR for the corpus for the branch at https://github.com/dhobsong/fuzz-corpora/tree/acert_corpus once this has been merged (or close to it), but I can do it sooner if needed.

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 30 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 61 days ago

JonathanWilbur commented:
Who is this PR waiting on?

dhobsong commented:
My understanding is that it is waiting on @openssl maintainer review. 
If that is not the case could  someone please let me know?

t8m reviewed (commented):
Can you please rebase this to resolve the conflicts? Also can you please run the fuzzer and create a pull request against the openssl/fuzz-corpora repo with some fuzz corpus?

I've preliminary reviewed this and I believe this could be merged.

dhobsong commented:
@t8m, I've rebased this PR, and created a new one for the fuzz corpus at https://github.com/openssl/fuzz-corpora/pull/23

I haven't updated the fuzz-corpora submodule commit in this series yet, to avoid breaking the git fetch in all the other CI tests.
I can add that in once the corpora is merged.

t8m reviewed (approved):
I think this PR is very well done according to the code style, it should not have negative impact on unrelated use-cases of OpenSSL and there is a fuzzer and tests. I am approving it.

JonathanWilbur commented:
@dhobsong It looks like you have conflicts on this now.

dhobsong commented:
@JonathanWilbur Thanks for the heads up.  Updated and pushed.

t8m reviewed (approved):
My approval holds

FdaSilvaYY commented on NEWS.md:
```diff
@@ -26,6 +26,8 @@ OpenSSL 3.3
   * Added exporter for CMake on Unix and Windows, alongside the pkg-config
     exporter.
 
+  * Added initial Attribute Cerficate (RFC 5755) support.
```
Typo : Cerficate

FdaSilvaYY commented:
> Would @FdaSilvaYY be willing to review now that all of his/her (I don't know) comments are addressed?

My approval don't count as I'm not part of project team/commiters.
@JonathanWilbur :  as an old grown guy (close to his half-century), i don't mind about pronoms ;)

dhobsong commented:
@FdaSilvaYY  Thank you again for the comment. 

I've fixed the typo and rebased again due to another merge conflict.

@t8m I see that you removed the "approval: otc review pending" label, but there is still an "approval: review pending" label on this PR. Does this need a second review?

t8m reviewed (approved):
Yes, a second review by an OpenSSL committer is still needed.

dhobsong commented:
Ok, thank you for the information, and for the review, of course.

slontis commented on fuzz/build.info:
```diff
@@ -193,6 +198,10 @@ IF[{- !$disabled{tests} -}]
   INCLUDE[decoder-test]=../include
   DEPEND[decoder-test]=../libcrypto
 
+  SOURCE[acert-test]=acert.c test-corpus.c fuzz_rand.c
```
Doesnt look like this requires either fuzz_rand.c or ../libssl

dhobsong replied:
Thank you.  Yes, neither are used in this test case.  Removed.

JonathanWilbur commented:
@slontis I don't know if you saw, but @dhobsong fixed the issue you pointed out.

JonathanWilbur commented:
@t-j-h Would it be rude for me to ask for a review from you for this one? This PR's been waiting over two years, and it's been reviewed a lot and already has one committer approval.

t-j-h commented:
> @t-j-h Would it be rude for me to ask for a review from you for this one? 

It is never rude to ask! I'm pretty swamped at the moment on other stuff but I'll add it to the list.

hlandau commented on doc/man3/X509_ACERT_print_ex.pod:
```diff
@@ -0,0 +1,112 @@
+
+=head1 HISTORY
+
+X509_ACERT_print() and X509_ACERT_print_ex() were added in OpenSSL 3.2.
```
3.3 in all files

hlandau commented on doc/man3/X509_ACERT_get0_holder_baseCertId.pod:
```diff
@@ -0,0 +1,118 @@
+=head1 RETURN VALUES
+
+All I<set0>/I<set1> routines return 1 for success and 0 for failure.
+All I<get0> functions return a pointer to the objects inner structure. These
```
object's

hlandau commented on crypto/x509/t_acert.c:
```diff
@@ -0,0 +1,292 @@
+             * are ignored.
+             */
+            if (sk_GENERAL_NAME_num(holderEntities) >= 1)
+                entity = sk_GENERAL_NAME_value(holderEntities, 0);
```
I think it is problematic to have a pretty-printer which doesn't show all of the information and just silently hides it.

JonathanWilbur replied:
Come to think of it, I also wonder what would happen if this received an empty SEQUENCE. Segfault?

hlandau commented on crypto/x509/x_ietfatt.c:
```diff
@@ -0,0 +1,189 @@
+ *
+ * Section 4.4.2 states that all values in the sequence MUST use the
+ * same choice of value (octet, oid or string).  This restriction is
+ * not explicitly enforced by the current implementation.
```
IMO, this should be enforced.

dhobsong replied:
@hlandau I think that it's easy enough to enforce this from the openssl API, but what about for importing these attributes from certificates generated elsewhere?  Do we need a  `d2i_OSSL_IETF_ATTR_SYNTAX()` implementation that returns `NULL` if the sequence value choices don't match?

dhobsong replied:
I've implemented this as I described here.

dhobsong replied:
@hlandau Did you have any comments about how the IetfAttrSyntax enforcement has been addressed here?

hlandau commented on test/certs/acert.pem:
```diff
@@ -0,0 +1,15 @@
+enYBmnH4rvW+SX8v5MzfGOu8VtHAKKBEcCbs2qGwYBEUUsCA/LYKSnOwn655wsd9
+k6KifIGOCVS4ZbK9pDyYcH/6/OmD1eSUwDZsd8CrJYDC31gAu7xqhOsBzR9DL3oD
+4Q==
+-----END ATTRIBUTE CERTIFICATE-----
```
What is the providence of this test certificate? Are there existing implementations whose output, or own test files, we can add to the test suite?

dhobsong replied:
This was created by me using bouncycastle. I will check again to see if I can find any test files from other implementations.

dhobsong replied:
Is there any advice on how to handle the licensing of PEM file test data?
For example, I have found a test certificate in the bouncycastle source, licensed under MIT.  Is it sufficient to add something like the following header to the PEM file?

```
Extracted from bouncycastle test case
Source: pkix/src/test/j2me/org/bouncycastle/cert/test/AttrCertTest.java
Copyright (c) 2000-2023 The Legion of the Bouncy Castle Inc. (https://www.bouncycastle.org)
SPDX-License-Identifier: MIT

-----BEGIN ATTRIBUTE CERTIFICATE-----
...
```

t8m replied:
This was always a problem. Unfortunately the CLA has to be applicable to the test data as well so IMO you cannot use this. Of course you can use certs you generated yourself by some 3rd party software.

dhobsong replied:
Ah yes, of course.  

NSA Cyber has a tool to generate platform certificates (https://github.com/nsacyber/paccor).  I'll try to generate something with that.

dghgit replied:
Hi, we've been sent a copy of the CLA and I'm in the process of signing it. You're welcome to make use of any test data we have published in the meanwhile (and anything else for that matter).

David Hook
Secretary
Legion of the Bouncy Castle Inc.
office@bouncycastle.org

dhobsong replied:
Oh, that's amazing.  Thank you very much.

@t8m, assuming that there is a CLA for bouncycastle on file, is the above header ok?  Should I add an "Acked-by" or something else to the commit message?  In that case, I think it makes sense to split this test data into a separate patch.

dghgit replied:
It was just emailed to the legal address.

t-j-h replied:
We have received it - it will be processed shortly.
Thanks @dghgit for sorting this out as it really helps to have good test data.

dhobsong replied:
I've added two certificates from bouncycastle test cases to the MR as a separate patch.
It seems to have been flagged, pending the CLA processing.

mattcaswell replied:
> It seems to have been flagged, pending the CLA processing.

The cla bot is complaining because you seem to have used two different author email addresses (both of them seem to be you) - but only one of them is in the CLA database.

dhobsong replied:
Thank you. I'll fix that, and I noticed that I have some CI failures too, so I will address those as well and re-push.

tatac1 replied:
@dhobsong 
I made some platform certificates (attribute certificates). I am happy if it helps.
please find the attach file, [acerts.zip](https://github.com/openssl/openssl/files/14414767/acerts.zip)

If there is a proper method of contribution other than attaching files, I would appreciate your guidance.

dhobsong replied:
@tatac1 Thank you for the certificates.  

If you want to contribute these, I think that you will also need to submit a [CLA](https://www.openssl.org/policies/cla.html), as it says in the CONTRIBUTING.md file.

This MR now already includes one TPM platform certificate, with similar fields to the certificates you provided, so that may be enough for now.  Your certificates use more attribute types, so should be very useful in future as support for more types of attributes are added.

tatac1 replied:
> @tatac1 Thank you for the certificates.
> 
> If you want to contribute these, I think that you will also need to submit a [CLA](https://www.openssl.org/policies/cla.html), as it says in the CONTRIBUTING.md file.
> 
> This MR now already includes one TPM platform certificate, with similar fields to the certificates you provided, so that may be enough for now. Your certificates use more attribute types, so should be very useful in future as support for more types of attributes are added.

I have signed an Individual Contributor License Agreement and sent it.

dhobsong replied:
> I have signed an Individual Contributor License Agreement and sent it.

Good to know that there are more certificates available if needed.

nhorman commented on crypto/x509/x509_acert.c:
```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright 2021 The OpenSSL Project Authors. All Rights Reserved.
```
Copyright should be 2023 here no?

nhorman commented on crypto/x509/x509_acert.h:
```diff
@@ -0,0 +1,22 @@
+/*
+ * Copyright 2021 The OpenSSL Project Authors. All Rights Reserved.
```
copyright 2024 also here and any in any other new files

dhobsong replied:
What is the convention here, because this PR has been open (and the files "available") since 2021.  Is the merge date used as the publication date?

mattcaswell replied:
That makes sense to me

dhobsong replied:
Ok great, thank you all for the comments. I'll leave this part as-is.

dhobsong commented:
I'm pushing this once more with the following changes:

- Rework the attribute certificate test to make it easier add new tests for parsing and printing of certificate data
- Pretty print all holder entity names in a certificate
- Enforce restriction that all entries in IetfAttrSyntax sequence use the same choice (both through API and import)
- Update man page "function added in" clauses to indicate OpenSSL 3.3
- Add missing man page history sections
- Add test data from bouncycastle
- Add TPM platform certificate test data
- Add more tests of IetfAttrSyntax parsing
- Pretty print Attribute Certificate serial numbers in hex, not decimal
- Rebase to TOT

tatac1 commented:
> I'm pushing this once more with the following changes:
> 
> * Rework the attribute certificate test to make it easier add new tests for parsing and printing of certificate data
> * Pretty print all holder entity names in a certificate
> * Enforce restriction that all entries in IetfAttrSyntax sequence use the same choice (both through API and import)
> * Update man page "function added in" clauses to indicate OpenSSL 3.3
> * Add missing man page history sections
> * Add test data from bouncycastle
> * Add TPM platform certificate test data
> * Add more tests of IetfAttrSyntax parsing
> * Pretty print Attribute Certificate serial numbers in hex, not decimal
> * Rebase to TOT

I received an email from OpenSSL with the following message:

```
You have been added to our CLA database and should now be able to contribute.
If you have any open pull requests, you must close and reopen them to remove
the "needs CLA" tag.
````

Consequently, the "[hold: cla required](https://github.com/openssl/openssl/labels/hold%3A%20cla%20required)" label should be removed, as this pertains to my submission of TPM platform certificate test data.

nhorman commented:
Is the ubsan failure relevant here?

dhobsong commented:
> Is the ubsan failure relevant here?

No, this failure seems unrelated.  
I've rebased and re-pushed and the failure is now gone.

dhobsong commented:
@nhorman Thank you for the approval.

FdaSilvaYY commented on CHANGES.md:
```diff
@@ -168,6 +168,12 @@ OpenSSL 3.3
 
    *Alexandr Nedvedicky*
 
+ * Added Attribute Certificate (RFC 5755) support. Attribute
```
This needs to move up to `OpenSSL 3.4` paragraph, while rebasing to  last master who recently jump to 3.4

FdaSilvaYY commented on util/libcrypto.num:
```diff
@@ -5548,3 +5548,86 @@ X509_STORE_get1_objects                 ?	3_3_0	EXIST::FUNCTION:
 OPENSSL_LH_set_thunks                   ?	3_3_0	EXIST::FUNCTION:
 OPENSSL_LH_doall_arg_thunk              ?	3_3_0	EXIST::FUNCTION:
 OSSL_HTTP_REQ_CTX_set_max_response_hdr_lines ?	3_3_0	EXIST::FUNCTION:HTTP
+PEM_read_X509_ACERT                     ?	3_3_0	EXIST::FUNCTION:STDIO
```
needs a `make update` while rebasing to Master / 3.4

FdaSilvaYY commented on test/x509_acert_test.c:
```diff
@@ -0,0 +1,174 @@
+    int valid;
+};
+
+static struct ietf_type_test_data ietf_syntax_tests[] = {
```
static const

FdaSilvaYY commented on test/x509_acert_test.c:
```diff
@@ -0,0 +1,174 @@
+}
+
+/* IetfAttrSyntax structure with one value */
+static unsigned char attr_syntax_single[] = {
```
static const ? 
No sure compiler will agree.

FdaSilvaYY commented on doc/man3/X509_get_serialNumber.pod:
```diff
@@ -59,6 +73,8 @@ L<X509_verify_cert(3)>
 available in all versions of OpenSSL.
 The X509_get0_serialNumber() function was added in OpenSSL 1.1.0.
+The X509_ACERT_get0_serialNumber() and X509_ACERT_set1_serialNumber()
+functions were added in OpenSSL 3.3.
```
OpenSSL 3.4

FdaSilvaYY commented on doc/man3/X509_get0_uids.pod:
```diff
@@ -45,6 +55,13 @@ L<X509_sign(3)>,
+
+X509_get0_uids() was added in OpenSSL 1.1.0.
+
+X509_ACERT_get0_issuerUID() was added in OpenSSL 3.3.
```
OpenSSL 3.4.
And so on.

FdaSilvaYY commented on doc/man3/X509_get0_uids.pod:
```diff
@@ -45,6 +55,13 @@ L<X509_sign(3)>,
 L<X509_verify_cert(3)>
 
+=head1 HISTORY
+
```
IS One blank line enough here ?

FdaSilvaYY commented on NEWS.md:
```diff
@@ -26,6 +26,8 @@ OpenSSL 3.3
   * Added exporter for CMake on Unix and Windows, alongside the pkg-config
     exporter.
 
+  * Added initial Attribute Certificate (RFC 5755) support.
```
needs to move up ...

FdaSilvaYY commented on doc/man3/OSSL_IETF_ATTR_SYNTAX.pod:
```diff
@@ -0,0 +1,97 @@
+
+OSSL_IETF_ATTR_SYNTAX_get0_policyAuthority(), OSSL_IETF_ATTR_SYNTAX_set0_policyAuthority(),
+OSSL_IETF_ATTR_SYNTAX_get_value_num(), OSSL_IETF_ATTR_SYNTAX_get0_value(), and
+OSSL_IETF_ATTR_SYNTAX_add1_value() were added in OpenSSL 3.3.
```
OpenSSL 3.4

FdaSilvaYY commented on doc/man3/X509_ACERT_add1_attr.pod:
```diff
@@ -0,0 +1,67 @@
+=head1 HISTORY
+
+X509_ACERT_add1_attr(), X509_ACERT_add1_attr_by_NID(), X509_ACERT_add1_attr_by_OBJ(),
+X509_ACERT_add1_attr_by_txt() and X509_ACERT_delete_attr() were added in OpenSSL 3.3.
```
3.4

FdaSilvaYY commented on doc/man3/X509_sign.pod:
```diff
@@ -68,6 +75,9 @@ available in all versions of OpenSSL.
 and X509_CRL_sign_ctx() functions were added in OpenSSL 1.0.1.
 
+The X509_ACERT_sign() and X509_ACERT_sign_ctx() functions were added
+in OpenSSL 3.3.
```
3.4

FdaSilvaYY commented on doc/man3/X509_verify.pod:
```diff
@@ -71,6 +75,8 @@ functions are available in all versions of OpenSSL.
 
 X509_REQ_verify_ex(), and X509_self_signed() were added in OpenSSL 3.0.
 
+X509_ACERT_verify() was added in OpenSSL 3.3.
```
3.4

FdaSilvaYY commented on doc/man3/OSSL_IETF_ATTR_SYNTAX_print.pod:
```diff
@@ -0,0 +1,41 @@
+
+=head1 HISTORY
+
+OSSL_IETF_ATTR_SYNTAX_print() was added in OpenSSL 3.3.
```
3.4

FdaSilvaYY commented on doc/man3/X509_ACERT_add_attr_nconf.pod:
```diff
@@ -0,0 +1,63 @@
+
+=head1 HISTORY
+
+The function X509_ACERT_add_attr_nconf() was added in OpenSSL 3.3.
```
3.4

FdaSilvaYY reviewed (commented):
Hope this PR will be merged soon.

t8m commented:
@dhobsong Please rebase this targetting 3.4 and update the fuzz-corpora submodule to the latest commit (I've merged the acert fuzz corpus).

dhobsong commented:
@t8m Will do.  Thank you.

dhobsong commented:
@t8m I've rebased to TOT and updated the fuzz/corpora submod reference.  Feels nice to see all tests, including fuzzing, passing.

@FdaSilvaYY Thank you once again for the review.

mattcaswell commented:
@nhorman - please reconfirm your approval

nhorman commented:
Approved

openssl-machine commented:
This pull request is ready to merge

mattcaswell commented:
Pushed!!!

I had to resolve some trivial conflicts during the merge.

Thanks @dhobsong for your contribution and your persistence in getting this PR in.

beldmit commented:
Nice success story!

dhobsong commented:
Great news!! Thank you so much to everyone who helped get this over the line. 
Especially @FdaSilvaYY for all the reviews, and @JonathanWilbur for keeping this moving and all the @openssl maintainers for the guidance and approvals at multiple stages.
